File: src/api/v1/endpoints/versions.py
"""Document versioning API endpoints.

Provides endpoints for:
- Version history listing
- Version content retrieval
- Diff viewing
- Rollback functionality
- Audit log access
"""

from typing import Optional, List
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel, Field

from src.api.deps import get_current_user, get_db
from src.db.models.user import User
from src.db.models.document_version import VersionType
from src.services.document_version_service import get_document_version_service

router = APIRouter(prefix="/documents/{document_id}/versions", tags=["versions"])


# Request/Response Models
class VersionInfo(BaseModel):
    id: str
    version_number: int
    version_type: str
    title: str
    content_hash: Optional[str]
    file_size: Optional[int]
    change_summary: Optional[str]
    changed_by: Optional[str]
    created_at: str
    diff_stats: Optional[dict]


class VersionHistoryResponse(BaseModel):
    document_id: str
    total_versions: int
    versions: List[VersionInfo]


class VersionContentResponse(BaseModel):
    document_id: str
    version_number: int
    title: str
    content: Optional[str]
    created_at: str


class DiffResponse(BaseModel):
    from_version: int
    to_version: int
    from_version_info: dict
    to_version_info: dict
    diff_content: Optional[str]


class CompareResponse(BaseModel):
    version_a: dict
    version_b: dict
    diff: Optional[str]
    same_content: bool


class RollbackRequest(BaseModel):
    reason: Optional[str] = Field(None, max_length=500)


class RollbackResponse(BaseModel):
    success: bool
    message: str
    new_version_number: int


class AuditLogEntry(BaseModel):
    id: str
    action: str
    action_details: Optional[dict]
    user_id: Optional[str]
    version_id: Optional[str]
    ip_address: Optional[str]
    success: bool
    error_message: Optional[str]
    created_at: str


class AuditLogResponse(BaseModel):
    document_id: str
    entries: List[AuditLogEntry]


# Helper function to get client info
def get_client_info(request: Request) -> tuple:
    """Extract client IP and user agent from request."""
    ip_address = request.client.host if request.client else None
    user_agent = request.headers.get("user-agent")
    return ip_address, user_agent


@router.get("", response_model=VersionHistoryResponse)
async def get_version_history(
    document_id: uuid.UUID,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> VersionHistoryResponse:
    """Get version history for a document.
    
    Returns a paginated list of all versions with metadata.
    """
    service = get_document_version_service(db)
    
    try:
        versions = await service.get_version_history(document_id, skip, limit)
        
        # Get total count
        total = len(versions)  # Could optimize with separate count query
        
        return VersionHistoryResponse(
            document_id=str(document_id),
            total_versions=total,
            versions=[VersionInfo(**v) for v in versions],
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{version_number}", response_model=VersionContentResponse)
async def get_version_content(
    document_id: uuid.UUID,
    version_number: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> VersionContentResponse:
    """Get content of a specific version.
    
    Returns the full content of the document at the specified version.
    """
    service = get_document_version_service(db)
    
    version = await service.get_version(document_id, version_number)
    if not version:
        raise HTTPException(
            status_code=404,
            detail=f"Version {version_number} not found for document {document_id}"
        )
    
    return VersionContentResponse(
        document_id=str(document_id),
        version_number=version.version_number,
        title=version.title,
        content=version.content,
        created_at=version.created_at.isoformat(),
    )


@router.get("/{from_version}/diff", response_model=DiffResponse)
async def get_version_diff(
    document_id: uuid.UUID,
    from_version: int,
    to_version: Optional[int] = Query(default=None, description="Target version (default: latest)"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> DiffResponse:
    """Get diff between two versions.
    
    If to_version is not specified, compares with the latest version.
    """
    service = get_document_version_service(db)
    
    try:
        diff = await service.get_diff(document_id, from_version, to_version)
        return DiffResponse(**diff)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{version_a}/compare/{version_b}", response_model=CompareResponse)
async def compare_versions(
    document_id: uuid.UUID,
    version_a: int,
    version_b: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CompareResponse:
    """Compare two specific versions side by side.
    
    Returns both version contents and the diff between them.
    """
    service = get_document_version_service(db)
    
    try:
        comparison = await service.compare_versions(document_id, version_a, version_b)
        return CompareResponse(**comparison)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/{version_number}/rollback", response_model=RollbackResponse)
async def rollback_to_version(
    document_id: uuid.UUID,
    version_number: int,
    request: Request,
    rollback_request: Optional[RollbackRequest] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> RollbackResponse:
    """Rollback document to a previous version.
    
    Creates a new version with the content from the target version.
    The current state is preserved as a version before rollback.
    """
    service = get_document_version_service(db)
    ip_address, user_agent = get_client_info(request)
    
    reason = rollback_request.reason if rollback_request else None
    
    try:
        document = await service.rollback(
            document_id=document_id,
            target_version=version_number,
            user_id=current_user.id,
            reason=reason,
            ip_address=ip_address,
            user_agent=user_agent,
        )
        
        # Get new version number
        versions = await service.get_version_history(document_id, skip=0, limit=1)
        new_version = versions[0]["version_number"] if versions else 0
        
        await db.commit()
        
        return RollbackResponse(
            success=True,
            message=f"Successfully rolled back to version {version_number}",
            new_version_number=new_version,
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/audit-log", response_model=AuditLogResponse)
async def get_audit_log(
    document_id: uuid.UUID,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=500),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AuditLogResponse:
    """Get audit log for a document.
    
    Returns all recorded actions performed on the document.
    """
    service = get_document_version_service(db)
    
    logs = await service.get_audit_log(document_id, skip, limit)
    
    return AuditLogResponse(
        document_id=str(document_id),
        entries=[AuditLogEntry(**log) for log in logs],
    )


# Additional endpoint for creating manual versions
class CreateVersionRequest(BaseModel):
    change_summary: Optional[str] = Field(None, max_length=500)


class CreateVersionResponse(BaseModel):
    version_id: str
    version_number: int
    created_at: str


@router.post("", response_model=CreateVersionResponse)
async def create_version(
    document_id: uuid.UUID,
    request: Request,
    create_request: Optional[CreateVersionRequest] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CreateVersionResponse:
    """Manually create a version snapshot.
    
    Creates a new version of the document's current state.
    Useful for marking important checkpoints.
    """
    service = get_document_version_service(db)
    ip_address, user_agent = get_client_info(request)
    
    change_summary = create_request.change_summary if create_request else None
    
    try:
        version = await service.create_version(
            document_id=document_id,
            version_type=VersionType.UPDATE,
            change_summary=change_summary,
            user_id=current_user.id,
            ip_address=ip_address,
            user_agent=user_agent,
        )
        
        await db.commit()
        
        return CreateVersionResponse(
            version_id=str(version.id),
            version_number=version.version_number,
            created_at=version.created_at.isoformat(),
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/api/versions.ts
/**
 * Document versioning API client
 */

import apiClient from './client';

// Types
export interface VersionInfo {
  id: string;
  version_number: number;
  version_type: string;
  title: string;
  content_hash: string | null;
  file_size: number | null;
  change_summary: string | null;
  changed_by: string | null;
  created_at: string;
  diff_stats: {
    lines_added: number;
    lines_removed: number;
    lines_changed: number;
  } | null;
}

export interface VersionHistoryResponse {
  document_id: string;
  total_versions: number;
  versions: VersionInfo[];
}

export interface VersionContentResponse {
  document_id: string;
  version_number: number;
  title: string;
  content: string | null;
  created_at: string;
}

export interface DiffResponse {
  from_version: number;
  to_version: number;
  from_version_info: {
    title: string | null;
    created_at: string | null;
  };
  to_version_info: {
    title: string | null;
    created_at: string | null;
  };
  diff_content: string | null;
}

export interface CompareResponse {
  version_a: {
    number: number;
    title: string;
    content: string | null;
    content_hash: string | null;
    created_at: string;
    changed_by: string | null;
  };
  version_b: {
    number: number;
    title: string;
    content: string | null;
    content_hash: string | null;
    created_at: string;
    changed_by: string | null;
  };
  diff: string | null;
  same_content: boolean;
}

export interface RollbackResponse {
  success: boolean;
  message: string;
  new_version_number: number;
}

export interface AuditLogEntry {
  id: string;
  action: string;
  action_details: Record<string, unknown> | null;
  user_id: string | null;
  version_id: string | null;
  ip_address: string | null;
  success: boolean;
  error_message: string | null;
  created_at: string;
}

export interface AuditLogResponse {
  document_id: string;
  entries: AuditLogEntry[];
}

export interface CreateVersionResponse {
  version_id: string;
  version_number: number;
  created_at: string;
}

// API Functions

/**
 * Get version history for a document
 */
export async function getVersionHistory(
  documentId: string,
  skip: number = 0,
  limit: number = 50
): Promise<VersionHistoryResponse> {
  const response = await apiClient.get<VersionHistoryResponse>(
    `/documents/${documentId}/versions`,
    { params: { skip, limit } }
  );
  return response.data;
}

/**
 * Get content of a specific version
 */
export async function getVersionContent(
  documentId: string,
  versionNumber: number
): Promise<VersionContentResponse> {
  const response = await apiClient.get<VersionContentResponse>(
    `/documents/${documentId}/versions/${versionNumber}`
  );
  return response.data;
}

/**
 * Get diff between two versions
 */
export async function getVersionDiff(
  documentId: string,
  fromVersion: number,
  toVersion?: number
): Promise<DiffResponse> {
  const response = await apiClient.get<DiffResponse>(
    `/documents/${documentId}/versions/${fromVersion}/diff`,
    { params: toVersion ? { to_version: toVersion } : {} }
  );
  return response.data;
}

/**
 * Compare two versions side by side
 */
export async function compareVersions(
  documentId: string,
  versionA: number,
  versionB: number
): Promise<CompareResponse> {
  const response = await apiClient.get<CompareResponse>(
    `/documents/${documentId}/versions/${versionA}/compare/${versionB}`
  );
  return response.data;
}

/**
 * Rollback to a previous version
 */
export async function rollbackToVersion(
  documentId: string,
  versionNumber: number,
  reason?: string
): Promise<RollbackResponse> {
  const response = await apiClient.post<RollbackResponse>(
    `/documents/${documentId}/versions/${versionNumber}/rollback`,
    reason ? { reason } : {}
  );
  return response.data;
}

/**
 * Get audit log for a document
 */
export async function getAuditLog(
  documentId: string,
  skip: number = 0,
  limit: number = 100
): Promise<AuditLogResponse> {
  const response = await apiClient.get<AuditLogResponse>(
    `/documents/${documentId}/versions/audit-log`,
    { params: { skip, limit } }
  );
  return response.data;
}

/**
 * Create a manual version snapshot
 */
export async function createVersion(
  documentId: string,
  changeSummary?: string
): Promise<CreateVersionResponse> {
  const response = await apiClient.post<CreateVersionResponse>(
    `/documents/${documentId}/versions`,
    changeSummary ? { change_summary: changeSummary } : {}
  );
  return response.data;
}